src/lib.rs

```diff
@@ -1,28 +1,35 @@
-pub(crate) mod default;
-pub(crate) mod local;
-pub(crate) mod stdlib;
-pub use self::local::Local;
+#![feature(const_fn)]
+#![cfg_attr(all(not(feature = "std"), not(doctest)), no_std)]
 
-#[cfg(feature = "std")]
-pub(crate) mod thread;
-#[cfg(feature = "std")]
-pub use self::thread::ThreadLocal;
+#[cfg(not(feature = "std"))]
+extern crate alloc;
 
-pub use self::default::DefaultLocality;
+#[doc(hidden)]
+pub mod local;
+mod stdlib;
 
-use stdlib::{any, cell::UnsafeCell, fmt, marker::PhantomData};
+#[cfg(not(feature = "std"))]
+use stdlib::{cell::UnsafeCell, fmt, marker::PhantomData, sync::atomic::Ordering};
 
+#[cfg(not(feature = "std"))]
+pub use local::{reset, set_locality};
+
+#[cfg(not(feature = "std"))]
 #[derive(Eq, PartialEq, Copy, Clone)]
 pub struct Id {
     value: usize,
     _not_send: PhantomData<UnsafeCell<()>>,
 }
 
 /// A locality implementation.
+#[cfg(not(feature = "std"))]
 pub trait Locality {
+    const MAX_LOCALITIES: usize;
```

Comment:
upon further consideration, i think this should probably be a function rather than a constant, since it might be `num_cpus::get()` or something. there should probably be the contract that it should always return the same value, though...